dsl_parser/elements/data_types.py

```diff
@@ -0,0 +1,203 @@
+########
+# Copyright (c) 2013 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+import copy
+
+from dsl_parser import constants
+from dsl_parser import elements
+from dsl_parser import exceptions
+from dsl_parser import utils
+from dsl_parser.elements import types, version as _version
+from dsl_parser.framework.elements import (
+    Element,
+    Dict,
+    DictElement,
+    Leaf,
+    StringElement)
+from dsl_parser.framework.requirements import (
+    Value,
+    Requirement,
+    sibling_predicate)
+
+
+class SchemaPropertyDescription(Element):
+
+    schema = Leaf(type=str)
+
+
+class SchemaPropertyType(Element):
+
+    schema = Leaf(type=str)
+
+    # requires will be modified later.
+    requires = {}
+
+    provides = ['component_types']
+
+    def validate(self, data_type, component_types):
+        if self.initial_value and self.initial_value not in \
+                constants.USER_PRIMITIVE_TYPES and not data_type:
+            raise exceptions.DSLParsingLogicException(
+                exceptions.ERROR_UNKNOWN_TYPE,
+                "Illegal type name '{0}'".format(self.initial_value))
+
+    def calculate_provided(self, data_type, component_types):
+        component_types = component_types or {}
+        if self.value and self.value not in constants.USER_PRIMITIVE_TYPES:
+            component_types = copy.copy(component_types)
```

Comment:
no need for `copy` as all the values injected are already copies of the original
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/elements.py#L107
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/elements.py#L122

https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/parser.py#L371
https://github.com/simon-bar/cloudify-dsl-parser/blob/dsl_parser_datatypes/dsl_parser/framework/parser.py#L387